schema-layer/data-structures/hashmap.md

```diff
@@ -0,0 +1,292 @@
+* `bitWidth`: The number of bits to use at each level of the data structure for determining the index of the entry or a link to the next level of the data structure to continue searching. The equation `2`<sup>`bitWidth`</sup> yields the arity of the HashMap nodes, i.e. the number of storage locations for buckets and/or links to child nodes.
+* `bucketSize`: The maximum array size of entry storage buckets such that exceeding `bucketSize` causes the creation of a new child node to replace entry storage.
+
+### Node properties
+
+Each node in a HashMap data structure contains:
+
+* `data`: An array, with a length of one to `2`<sup>`bitWidth`</sup>.
+* `map`: A bitfield where the first `2`<sup>`bitWidth`</sup> bits are used to indicate whether a bucket or child node link is present at each possible index of the node.
+
+An important property of a HAMT is that the `data` array only contains active elements. Indexes in a node that do not contain any values (in buckets or links to child nodes) are not stored and the `map` bitfield is used to determine the `data` whether values are present and the array index of present values using a [`popcount()`](https://en.wikipedia.org/wiki/Hamming_weight). This allows us to store a maximally compacted `data` array for each node.
+
+### Schema
+
+The **root block** of an IPLD HashMap contains the same properties as all other blocks, in addition to configuration data that dictates how the algorithm below traverses and mutates the data structure.
+
+See [IPLD Schemas](/schema-layer/schemas) for a definition of this format.
+
+```ipldsch
+# Root node layout
+type HashMapRoot struct {
+  hashAlg String
+  bitWidth Int
+  bucketSize Int
+  map Bytes
+  data [ Element ]
+}
+
+# Non-root node layout
+type HashMapNode struct {
+  map Bytes
+  data [ Element ]
+}
+
+type Element union {
+  | Link "link"
+  | Bucket "bucket"
+} representation keyed
+
+type Bucket list [ BucketEntry ]
+
+type BucketEntry struct {
+  key Bytes
+  value Value
+} representation tuple
+
+type Value union {
+  | Bool bool
+  | String string
+  | Bytes bytes
+  | Int int
+  | Float float
+  | Map map
+  | List list
+  | Link link
+} representation kinded
+```
+
+Notes:
+
+* `hashAlg` in the root block is a string name that corresponds to a [multihash](https://github.com/multiformats/multihash) as found in the [multiformats table](https://github.com/multiformats/multicodec/blob/master/table.csv).
+* `bitWidth` in the root block must be at least `2` and no more than `12`. ***(TODO: this is arbitrary and currently IAMap only supports a maximum of `8` and go-hamt-ipld _only_ supports `8`. `12` yields a maximum `data` array length of `4096`, is it reasonable to go higher? A two-byte maximum would be neat, but `65,536` elements is maybe too much.)***
+* `bucketSize` in the root block must be at least `1`. ***(TODO: does it need a maximum?)***
+* Values can be any kind but cannot be `Null`
```

Comment:
+1 not supporting null will be problematic; the usual understanding in the data model is that a map can contain null values.

Schemas flip the default (nullability must be opt-in), but even there it would be a little startling to have some advanced layouts which aren't allowed to be used if the nullable modifier is on the value type.

Reply:
OK, I can live with that. I've built `null` as a signal into a lot of JS implementation but that should probably have been `undefined`.

The schema can probably become:

```
type BucketEntry struct {
  key Bytes
  value Value (implicit Null)
} representation tuple
```

Reply:
I’ve taken a note to add an explicit “not found” message from `get` functions in IPLD Generics.

Reply:
We should avoid making decisions in IPLD because $lang.

Reply:
> We should avoid making decisions in IPLD because $lang.

A mindset I'm still adapting to tbh. Happy to be kept in check on it of course.

I've removed this limitation so values can now be `Null` so the schema for an entry is:

```ipldsch
type BucketEntry struct {
  key Bytes
  value Value (implicit "null")
} representation tuple
```

which looks like `[ "akey": "avalue" ]` for a standard key/value pair, and `[ "anull" ]` for a key/null pair.

The docs also had 2 other places that mentioned returning `Null` if the value wasn't found, I've replaced those with: "return an empty value (as appropriate for the implementation platform)".